File: 653 - Two Sum IV - Input is a BST.java
import java.util.HashSet;
import java.util.Set;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    Set<Integer> s = new HashSet<>();
    int n;
    public boolean findTarget(TreeNode root, int k) {
        s.clear();
        this.n = k;
        return dfs(root);
    }
    public boolean dfs(TreeNode root) {
        if (root == null) {
            return false;
        }
        boolean l, r, m;
        l = dfs(root.left);
        m = s.contains(this.n - root.val);
        s.add(root.val);
        r = dfs(root.right);
        return l || r || m;
    }
}

File: 653 - Two Sum IV - Input is a BST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    s = set()
    def findTarget(self, root: TreeNode, k: int) -> bool:
        self.s.clear()
        return self.dfs(root, k)

    def dfs(self, root, k):
        if root is None:
            return False
        l = self.dfs(root.left, k)
        m = (k - root.val) in self.s
        self.s.add(root.val)
        r = self.dfs(root.right, k)
        return l or r or m